core/src/engine/single_consumer_event.cpp

```diff
@@ -56,34 +51,34 @@ bool SingleConsumerEvent::WaitForEventUntil(Deadline deadline) {
   impl::TaskContext& current = current_task::GetCurrentTaskContext();
   if (current.ShouldCancel()) return GetIsSignaled();
 
-  LOG_TRACE() << "WaitForEvent()";
-  impl::WaitListLight::SingleUserGuard guard(*lock_waiters_);
-  EventWaitStrategy wait_manager(*lock_waiters_, is_signaled_, current,
-                                 deadline);
+  LOG_TRACE() << "WaitForEventUntil()";
+  EventWaitStrategy wait_manager(*this, current, deadline);
 
-  bool was_signaled = false;
-  while (!(was_signaled = GetIsSignaled()) && !current.ShouldCancel()) {
```

Comment:
`ShouldCancel` check is already done by `Sleep`, there is no need to do it twice